src/v5/attached.rs

```diff
@@ -0,0 +1,143 @@
+use bytes::{Buf, Bytes};
+use protobuf::Message;
+
+use crate::{aes::IvAndCiphertext, icl_header_v4::V4DocumentHeader, Error};
+
+use super::key_id_header::{self, KeyIdHeader};
+
+type Result<A> = std::result::Result<A, Error>;
+
+#[derive(Debug, PartialEq)]
+pub struct AttachedDocument {
+    pub key_id_header: KeyIdHeader,
+    pub edek: V4DocumentHeader,
+    pub edoc: IvAndCiphertext,
+}
+
+/// Construct an IronCore attached EDOC from the constituent parts.
+pub fn encode_attached_edoc(
+    AttachedDocument {
+        key_id_header,
+        edek,
+        edoc,
+    }: &AttachedDocument,
+) -> Result<Bytes> {
+    let key_id_header_bytes = key_id_header.write_to_bytes();
+    let encoded_edek = edek.write_to_bytes().expect("Writing to bytes is safe");
+    if encoded_edek.len() > u16::MAX as usize {
+        Err(Error::HeaderLengthOverflow(encoded_edek.len() as u64))
+    } else {
+        let len = encoded_edek.len() as u16;
+
+        let result = [
+            key_id_header_bytes.as_ref(),
+            &len.to_be_bytes(),
+            &encoded_edek,
+            &edoc.0, // Note that the edoc is written without a header since it's an attached document.
+        ]
+        .concat();
+        Ok(result.into())
+    }
+}
+
+/// Breaks apart an attached edoc into its parts.
+pub fn decode_attached_edoc(b: Bytes) -> Result<AttachedDocument> {
+    let (key_id_header, mut attached_document_with_header) =
+        key_id_header::decode_version_prefixed_value(b)?;
+    if attached_document_with_header.len() > 2 {
+        let len = attached_document_with_header.get_u16();
```

Comment:
1. Maybe `header_len` as a name to indicate that it's just the V4DocumentHeader length
2. Can you comment here that this advances the byte buffer by 2 bytes? I forgot that and was confused for a bit (or byte)